doc/fluid/design/data_pipeline/data_pipeline.md

```diff
@@ -0,0 +1,92 @@
+# Design Doc: Fluid Data Pipeline
+
+This document is about how Fluid training and inference programs read data.
+
+## Standard Data Format
+
+### Case 1: Data from Files
+
+Consider a Fluid training program, `resnet50.py`, needs to read data from disk:
+
+```bash
+cat data | python resnet50.py
+```
+
+Since the person who collects the data might be different from the person who wrote resnet50.py:
+
+1. Fluid operators used in `resnet50.py` can recognize the file format of `data`, or, we need a standard data format.
+1. These operators need to be able to read from the standard input.
+
+### Case 2: Data from Online Generators
+
+Instead of files, data might come online.  For example:
+
+- Data generator for performance benchmarking.
+- Data generator for training special models like [GAN](https://en.wikipedia.org/wiki/Generative_adversarial_network).
+- The online data stream in production systems, e.g., online advertising.
+
+Consider that 
+
+1. data generators could crash and be restarted (by Kubernetes or other cluster management systems), and
+1. the network/pipe connection between could the generator and the trainer may break,
+
+we need
+
+1. the data format is fault-tolerable.
```

Comment:
We also need the fluid can read from multipe files concurrently? Suppose there are many data generators, they should generate the data into multiples files to avoid mutual interference.